lib/dinosaur.js

```diff
@@ -89,55 +87,76 @@ Dinosaur.prototype.move = function(floors) {
 
 Dinosaur.prototype.onAFloor = function(floors) {
   var dino = this;
-  var floor_receiver;
   var floor;
-  var dino_collider = {x: dino.x+dino.width/2, y: dino.y+dino.height};
   for(var i = 0; i< floors.length; i++) {
     floor = floors[i];
-    floor_receiver = {minX: floor.x,
-                      maxX: floor.x+floor.width,
-                      minY: floor.y,
-                      maxY: floor.y+floor.height};
-    if(Collision.collision(dino_collider, floor_receiver)) {
+    if (onThisFloor(floor, dino)) {
       return true;
     }
   }
   return false;
 };
 
 Dinosaur.prototype.jump = function(floors) {
-  if ( this.count < this.jumpSteps ){
-    this.count++;
-    if (this.y - this.jumpSize > 0){
-      this.y -= this.jumpSize;
-    } else {
-      this.y = 0;
-      this.count = this.jumpSteps;
-    }
-  } else if (this.count >= this.jumpSteps && this.count < 2*this.jumpSteps) {
-    this.count++;
-    var dino = this;
-    floors.forEach(function(floor){
-      var dino_collider = {x: dino.x+dino.width/2, y: dino.y+dino.height};
-      var floor_receiver = {minX: floor.x,
-                             maxX: floor.x+floor.width,
-                             minY: floor.y,
-                             maxY: floor.y+floor.height};
-      if(Collision.collision(dino_collider, floor_receiver)) {
-        dino.y = floor.y-floor.height/2-dino.height-2;
-        dino.status = null;
-        dino.count = 0;
-        return dino;
-      }
-    });
-    this.y += this.jumpSize;
-  } else if (this.count === 2*this.jumpSteps) {
-    this.status = null;
-    this.count = 0;
+  var dino = this;
+  if ( stillJumpingUp(dino) ){
+    dontHitCeiling(dino);
+  } else if (jumpingDown(dino)) {
+    findNearestFloor(floors, dino);
+  } else if (finishedJumpingAndFalling(dino)) {
+    resetDino(dino);
   }
-  return this;
+  return dino;
 };
 
+function onThisFloor(floor, dino) {
```

Comment:
hmmm... should we make these module functions or should we keep them as methods on the Dinosaur object?